### PR TITLE
add FResourceRequestEx interface

### PR DESCRIPTION
**Pull request by ZiyueHuang**
## Description ##
For https://github.com/apache/incubator-mxnet/issues/10509 and https://github.com/apache/incubator-mxnet/issues/9782.

For test, I have run matrix factorization with momentum sgd when `lazy_update=False` on GPU.

cc @eric-haibin-lin 

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/imperative/imperative_utils.h [hunk 1]
```diff
@@ -192,6 +192,7 @@ inline void SetDependency(const nnvm::NodeAttrs& attrs,
                    const DispatchMode dispatch_mode) {
   static auto& fmutate = nnvm::Op::GetAttr<nnvm::FMutateInputs>("FMutateInputs");
   static auto& ftmp_resource = nnvm::Op::GetAttr<FResourceRequest>("FResourceRequest");
+  static auto& ftmp_resourceEx = nnvm::Op::GetAttr<FResourceRequestEx>("FResourceRequestEx");
```
Comment (eric-haibin-lin) [inline src/imperative/imperative_utils.h]: ftmp_resourceEx -> ftmp_resource_ex

File: include/mxnet/op_attr_types.h [hunk 2]
```diff
@@ -219,6 +219,15 @@ using FStatefulComputeEx = std::function<void (const OpStatePtr& state,
  */
 using FResourceRequest = std::function<
   std::vector<ResourceRequest> (const NodeAttrs& n)>;
+/*!
+ * \brief The resource request from the operator for FComputeEx.
```
Comment (eric-haibin-lin) [inline include/mxnet/op_attr_types.h]: The resource request from the operator. (remove "for FComputeEx")
An operator could register ResourceRequestEx, or ResourceRequest, or neither.

File: src/executor/attach_op_resource_pass.cc [hunk 3]
```diff
@@ -47,8 +50,13 @@ Graph AttachOpResources(Graph g) {
...
     const auto op = inode.source->op();
-    if (fresource.count(op) != 0) {
-      auto reqs = fresource[op](inode.source->attrs);
+    const bool rsc_req = (fresource.count(op) != 0);
+    const bool rsc_ex_req = (fresource_ex.count(op) != 0);
+    if (rsc_req || rsc_ex_req) {
+      auto reqs = rsc_ex_req ? fresource_ex[op](inode.source->attrs,
+                                                dev_masks[nid],
+                                                vdispatch[nid])
+                             : fresource[op](inode.source->attrs);
```
Comment (eric-haibin-lin) [inline src/executor/attach_op_resource_pass.cc]: add CHECK(!(rsc_req && rsc_ex_req)) so that only one of them is registered

Comment (ZiyueHuang) [reply src/executor/attach_op_resource_pass.cc]: Added.

**Comment by ZiyueHuang** [comment]
any further comments? @eric-haibin-lin @zheng-da @piiswrong

**Pull request merged**